D8: Read xml layout

achaikou opened the pull request:
With this PR we can read embedded xml.
For now only works with default mapping. (Different bytes require more changes, different types I have not yet looked at).

- Looks like weird segfaults are defeated (I'll have to see if we could run our python tests with valgrind in CI too, it worked really well).
- I am sure there is still more things to prettify here, but my brain already cannot see them and is counting on yours.
- Special changes for Windows because one job started failing. Probably some update happened...

achaikou committed:
Expose stanza information to python

Note that it is not yet exposed to python user interface.

achaikou committed:
Move overwrite_mapping to a better location

Previous function location was a mistake. A lot of code related to trace
mapping layout handling is expected and it would clutter more important
functions.

achaikou committed:
Introduce concept of stanza

According to specification:
"The data in the Extended Textual File Header are organized in the form
of stanzas."

So instead of dealing directly with extended textual file header
records, we start dealing with stanzas.

Note that correct extended headers count when SEGY_BIN_EXT_HEADERS=-1
can be retrieved via stanzas.size() already. However all affected places
(ext_headers property in python or segy_trace0 calculations) are not yet
updated as support requires non-trivial changes.

achaikou committed:
Add name to segy_entry_definition

At the moment we have enough information to swiftly access all field
offsets needed internally in the library by name. This is what
segy_header_mapping.name_to_offset is for.

However as we now move to user-defined xml, having names for
predefined-only entries is not enough. We do not have a structure to
store names that are not part of SEG-Y standard.

Unless we create it, we would have to disregard custom name information
on parsing. So if some file has a non-standard field in the header, user
won't know its name.

To prevent this, add NULL-terminated name information to
segy_entry_definition. It is not needed by the library itself, but would
be used in data-presenting functions.

Default names are not added to binary_map/traceheader_default_map. It is
not yet clear if it is a good idea or not.

achaikou committed:
Exchange segy_field_data datatype with entry_type

Entry_type is a more general type than our internal datatype, so it
makes more sense to store it and avoid performing internal calculations
when it becomes needed.

It also fixes the bug, that for some reason only got discovered by
valgrind check after introduction of the next commit.

When we pass fields outside of the valid range to our C functions, we
might read from the incorrect memory place as we do not check the field
range when extracting data directly from the map. We do it only inside
inner get_field/set_field functions.

By storing entry_type in segy_field_data we avoid having to deal with
this altogether.

achaikou committed:
Use custom xml traceheader mapping layout

achaikou committed:
Comply with Windows CI recommendations

ajaust reviewed (commented):
I don't think I had any major remarks. I have some questions for clarification and some comments w.r.t. to C++ stuff. The C++ remarks are mostly ideas how one could do it differently, but you have to see if you see the benefit in doing it the suggested way.

achaikou committed:
Fix typo in offset_to_entry_definition

achaikou reviewed (commented):
Great! Thanks for the review! :cat:

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -2538,6 +2564,102 @@ extern "C" int segy_parse_layout_xml(
+        }
+    }
+
+    size_t headers = end_index - start_index;
```
First half of the `extract_xml_from_layout_stanza` is very had to follow and get an intuition for what we are tying to achieve and why. My best guess is that we are calculating the number of extended text headers used to define the xml.  As comments is not an option, could we move this into a function with a descriptive name?

yngve793 reviewed (approved):
Out of 12 comment is only seams to register 2 when writing this summary. Let us hope that Github works better behind the behind the frontend. 

All comments are written independent of comments from @ajaust, thus there might be repeated comments or issues that are already resolved in fixups. 

I struggle a bit to follow the over all thought in the c++ and c part of the code. Overall it looks fine, but confidence is at medium. Creating a propper suggestion will imply solving the problem all over again. 

Hopefully some of the comments will be useful.

achaikou reviewed (commented):
Thanks for the review!
(And of course it was useful! :cat: )

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -600,6 +623,9 @@ PyObject* segyopen( segyfd* self, PyObject* args, PyObject* kwargs ) {
     int err = segy_binheader( ds, binary );
     if( err ) return Error( err );
 
+    err = parse_extended_text_headers( self, binary );
```
Only the SEGY_BIN_EXT_HEADERS value is used. Why not pass the value it self over the pointer to the buffer. Tricks like this makes the code unnecessary hard to read.

achaikou reviewed (commented):
I added a few new commits, not as fixups for at least some clarity. Without proper rebase this PR is a mess anyway. Prettyfing will be thus done later.
Changes are done after conversation with PO and I think somewhere along the lines of your suggestions, Yngve.

ajaust reviewed (commented):
Thanks for the replies and explanations. To me it looks good overall. 🙂 

I marked most comments as resolved. The comments I did not mark as resolved usually had a (positive?) reaction, but I either didn't see any "Outdated" marker, or didn't find any code change related to the comment (I might have missed it). For some stuff, it could also be that the comment is not relevant anymore since the code disappeared due to moving to reading stanzas instead of extended textual headers.

I think I also added new 2-3 minor comments. I am curious

achaikou committed:
Define D8 xml parsing function

It seems a bit complicated to parse xml in C, we are not sure of the
best approach yet.

Parsing it in python is easy though. (If we had pybind11, it probably
would be easy to translate result to C/C++ as well).

So for now parsing implementation is implemented for python binding
directly, instead of being an integral part of the C library.

achaikou committed:
Drop encoding for extended textual headers

It appears that extended textual headers may contain binary information
as well.
Thus it is better to return information raw. It also appears very
unlikely that extended text header would be EBCDIC-encoded and thus need
encoding/decoding.

achaikou reviewed (commented):
Thanks!
Something is fixed.
Everything is rebased and split back into meaningful commits again.

yngve793 commented on lib/src/segy.c:
```diff
@@ -3033,3 +3033,142 @@ int segy_rotation_cw( segy_datasource* ds,
+}
+
+/** Reads (parts of) text header from datasource and parses stanza header. */
+static int read_stanza_header(
```
Consider to combine `read_stanza_header` and `segy_read_stanza_header` into `segy_read_stanza_header`.
Part of `read_stanza_header` is the same for both cases used in `segy_read_stanza_header`.

yngve793 reviewed (approved):
I am basically fine with it now. 
Have a look at the comments and adapt if you agree.

achaikou reviewed (commented):
Thanks for the review!

ajaust reviewed (approved):
Just went through everything again to look for unresolved comments. Looks all good to me. 🙂
